tracee-rules/engine/engine.go

```diff
@@ -62,24 +67,27 @@ func NewEngine(sigs []types.Signature, sources EventSources, output chan types.F
 	return &engine
 }
 
+// signatureStartFunc is the signature handling business logics.
+func signatureStartFunc(signature types.Signature, c chan types.Event) {
+	for e := range c {
+		if err := signature.OnEvent(e); err != nil {
+			meta, _ := signature.GetMetadata()
+			log.Printf("error handling event by signature %s: %v", meta.Name, err)
+		}
+	}
+}
+
 // Start starts processing events and detecting signatures
 // it runs continuously until stopped by the done channel
 // once done, it cleans all internal resources, which means the engine is not reusable
 // note that the input and output channels are created by the consumer and therefore are not closed
 func (engine *Engine) Start(done chan bool) {
 	go engine.consumeSources(done)
+	engine.signaturesMutex.RLock()
 	for s, c := range engine.signatures {
-		defer close(c)
-		go func(s types.Signature, c chan types.Event) {
-			for e := range c {
-				err := s.OnEvent(e)
-				if err != nil {
-					meta, _ := s.GetMetadata()
-					log.Printf("error handling event by signature %s: %v", meta.Name, err)
-				}
-			}
-		}(s, c)
+		go signatureStartFunc(s, c)
 	}
+	engine.signaturesMutex.RUnlock()
 	<-done
```

Comment:
Please rebase on main.
This code has changed

Reply:
done.